com.unity.perception/Runtime/GroundTruth/Labelers/KeypointLabeler.cs

```diff
@@ -86,33 +92,95 @@ protected override void Setup()
             m_KnownStatus = new Dictionary<uint, CachedData>();
+
+            m_AsyncAnnotations = new Dictionary<int, AsyncAnnotation>();
+            m_Keypoints = new Dictionary<int, Dictionary<uint, KeypointEntry>>();
+            m_ToReport = new List<KeypointEntry>();
+            m_CurrentFrame = 0;
+
+            perceptionCamera.InstanceSegmentationImageReadback += OnInstanceSegmentationImageReadback;
+        }
+
+        bool AreEqual(Color32 lhs, Color32 rhs)
+        {
+            return lhs.r == rhs.r && lhs.g == rhs.g && lhs.b == rhs.b && lhs.a == rhs.a;
+        }
+
+        void OnInstanceSegmentationImageReadback(int frameCount, NativeArray<Color32> data, RenderTexture renderTexture)
+        {
+            if (!m_AsyncAnnotations.TryGetValue(frameCount, out var asyncAnnotation))
+                return;
+
+            m_AsyncAnnotations.Remove(frameCount);
+
+            if (!m_Keypoints.TryGetValue(frameCount, out var keypoints))
+                return;
+
+            var width = renderTexture.width;
+
+            m_Keypoints.Remove(frameCount);
+
+            m_ToReport.Clear();
+
+            foreach (var keypointSet in keypoints)
+            {
+                if (InstanceIdToColorMapping.TryGetColorFromInstanceId(keypointSet.Key, out var idColor))
+                {
+                    var shouldReport = false;
+
+                    foreach (var keypoint in keypointSet.Value.keypoints)
+                    {
+                        // If the keypoint isn't mapped to a body part keep it at 0
+                        if (keypoint.state == 0) continue;
+
+                        if (keypoint.x < 0 || keypoint.x > width || keypoint.y < 0 || keypoint.y > renderTexture.height)
+                        {
+                            keypoint.state = 0;
+                            keypoint.x = 0;
+                            keypoint.y = 0;
+                        }
+                        else
+                        {
+                            // Get the pixel color at the keypoints location
+                            var height = renderTexture.height - (int)keypoint.y;
+                            var pixel = data[height * width + (int)keypoint.x];
+
+                            keypoint.state = AreEqual(pixel, idColor) ? 2 : 1;
+                            shouldReport = true;
+                        }
+                    }
+
+                    if (shouldReport)
+                        m_ToReport.Add(keypointSet.Value);
+                }
+            }
+
+            KeypointsComputed?.Invoke(frameCount, m_ToReport);
+            asyncAnnotation.ReportValues(m_ToReport);
         }
 
         /// <inheritdoc/>
         protected override void OnBeginRendering()
         {
-            var reporter = perceptionCamera.SensorHandle.ReportAnnotationAsync(m_AnnotationDefinition);
+            m_CurrentFrame = Time.frameCount;
+
+            m_Keypoints[m_CurrentFrame] = new Dictionary<uint, KeypointEntry>();
```

Comment:
For performance reasons we should avoid allocating managed objects on a per-frame basis. It's a death by a thousand cuts kind of issue. Give some thought to how you could use the Native* types by allocating all the memory you need up front. If you don't see a path, we can leave this exercise for another time.

Reply:
I looked around at this. I couldn't really find a way to do this without going fully into ECS and turning the class into a component. So I think we may have to punt on this for right now and come back to it.